### PR TITLE
joshfischer/next.js client

### PR DESCRIPTION
**Pull request by joshfischer1108**
This PR adds an authorization code flow library for Next.js. 

I want to get this out in the next release to get people using and testing the system, it's definitely not perfect by any means. 

I'm hoping this MVP will help more people get involved in the project. 

Any feedback, suggestions, or complaints 😄  are welcome.  Javascript/Typescript is not in my wheelhouse!

Docs and a Next.js example implementation are coming after this PR.

### CODE REVIEW

**Comment by joshfischer1108** [comment]
I tried to wrap `useEffect` (listed below) in one of the client methods but ran into issues.   So I left the call in the Next.js example component for now.

```
 useEffect(() => {
        fetch(iridiumClient.generateExchangeURL(), iridiumClient.generateExchangeHeaders())
            .then((response) => {
                const exchangeSuccessful = iridiumClient.processExchangeResponse(response)
                    .then((isSuccessful) => {
                        console.log('exchange successful', exchangeSuccessful)
                        if (isSuccessful) {
                            setDisplayMessage("The user has successfully authorized your application")
                        } else {
                            setDisplayMessage("There was an error authorizing your application")
                        }
                    })

            }).catch(error => {
                console.log("error fetching ", error)
        })
    }, []);

```

File: iridium-client-nextjs/src/lib/services/accessToken.service.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { CookieService } from './cookie.service';
+import { OauthConstants } from './oauth-constants';
+import { useEffect } from 'react';
```
Comment (klaurtar) [inline iridium-client-nextjs/src/lib/services/accessToken.service.ts:3]: This is not being used. I'd personally remove it.

Also, useEffect can only be used in a React function component. This is a component that returns some jsx. Typically with a file extension .tsx. A good rule of thumb is a .ts file should not be using hooks that have the word "use"

File: iridium-client-nextjs/src/lib/services/cookie.service.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+export class CookieService {
+    public getCookie(name: string) {
+        let ca: Array<string> = document.cookie.split(';');
+        let caLen: number = ca.length;
+        let cookieName = `${name}=`;
+        let c: string;
```
Comment (klaurtar) [inline iridium-client-nextjs/src/lib/services/cookie.service.ts:6]: The let keyword here typically lets others know that this variable is going to change somewhere down the line. These all look like they're assigned once and that is it. I'd change these to const, and try and make the naming a little more intuitive due to the open source nature of this project. I'm not sure what ca, c, caLen mean

File: iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { CookieService } from './cookie.service';
+import { OauthConstants } from './oauth-constants';
+import { exchange } from './accessToken.service';
+
+export class ExchangeURLGeneratorService {
+    generate() {
+        const queryString = window.location.search;
```
Comment (klaurtar) [inline iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts:7]: `const {search: queryString} = window.location`

Personally I would destructure out search from the object and alias it as queryString if you still want to use that name

File: iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    authorize() {
+        const queryString = window.location.search;
+        const urlParams = new URLSearchParams(queryString);
+        const cookieService = new CookieService();
+        const code = urlParams.get('code')
+        const state = urlParams.get("state")
+        if (code !== null && state === cookieService.getCookie(OauthConstants.STATE)) {
+            return exchange(code)
+        }
+        console.log('code was null or state didn\'t match')
```
Comment (klaurtar) [inline iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts:22]: Do you need this console.log statement here?

File: iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { CookieService } from './cookie.service';
+import { OauthConstants } from './oauth-constants';
+import { exchange } from './accessToken.service';
+
+export class ExchangeURLGeneratorService {
+    generate() {
+        const queryString = window.location.search;
+        const urlParams = new URLSearchParams(queryString);
+        const output = this.authorize();
+        return process.env.NEXT_PUBLIC_IRIDIUM_DOMAIN + 'oauth/token?grant_type=authorization_code&code=' + urlParams.get('code') + '&redirect_uri=' + output?.redirectUri + '&client_id=' + output?.clientId + '&code_verifier=' + output?.codeVerifier + '&state=' + output?.state;
```
Comment (klaurtar) [inline iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts:10]: String concat that is this long is nasty business. Personally I'd use a template literal so I don't have to worry about missing a plus and its a "little" more readable

``${
  process.env.NEXT_PUBLIC_IRIDIUM_DOMAIN
}oauth/token?grant_type=authorization_code&code=${urlParams.get(
  'code'
)}&redirect_uri=${output?.redirectUri}&client_id=${
  output?.clientId
}&code_verifier=${output?.codeVerifier}&state=${output?.state}`;`

Comment (klaurtar) [reply iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts:10]: `

`${process.env.NEXT_PUBLIC_IRIDIUM_DOMAIN }oauth/token?grant_type=authorization_code&code=${urlParams.get( 'code' )}&redirect_uri=${output?.redirectUri}&client_id=${ output?.clientId }&code_verifier=${output?.codeVerifier}&state=${output?.state}`;

`

File: iridium-client-nextjs/src/lib/services/identity.service.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+                .then((response) => response.json())
+                .then((res) => {
+                    if (res.status === 200) {
+                        this.cookieService.setCookie('iridium-token', res.access_token, 1, OauthConstants.COOKIE_PATH)
+                    } else {
+                      console.error('error!')
+                    }
+                })
+
+        }, [])
```
Comment (klaurtar) [inline iridium-client-nextjs/src/lib/services/identity.service.ts:26]: I have major questions about this piece of code. Is it working as expected? 

.Then chaining in a useEffect is not fun. Wrap it in an IIFE and async/await

`useEffect(() => {
  (async () => {
    const response = await fetch(
      process.env.NEXT_PUBLIC_IRIDIUM_DOMAIN + 'identities',
      {
        headers: {
          Accept: 'application/vnd.iridium.id.identity-response.1+json',
          Authorization: `Bearer ${token}`,
        },
        method: 'POST',
      }
    );
    const res = await response.json();

    if (res.status === 200) {
      this.cookieService.setCookie(
        'iridium-token',
        res.access_token,
        1,
        OauthConstants.COOKIE_PATH
      );
    } else {
      console.error('error!');
    }
  })();
}, []);`

File: iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts (hunk 3 above)
Comment (joshfischer1108) [reply iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts:7]: What value does this add?  Is it just the javascript normal way of doing things?  Or are there other benefits I'm missing?

File: iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts (hunk 4 above)
Comment (joshfischer1108) [reply iridium-client-nextjs/src/lib/services/exchange-URL-generator.service.ts:22]: Nope, I was sloppy.  Thank you.

File: iridium-client-nextjs/src/lib/services/identity.service.ts (hunk 6 above)
Comment (joshfischer1108) [reply iridium-client-nextjs/src/lib/services/identity.service.ts:26]: Good call.  I meant to remove this.  Thank you.

Comment (joshfischer1108) [reply iridium-client-nextjs/src/lib/services/identity.service.ts:26]: And no, it wasn't working 🙃

**Pull request merged**